Code-For-Chicago/Voma-backend
# Volunteers create endpoint

Adds in an end point for creating volunteers. Also adds in rspec and a spec for that end point. The rspec addition should help make it easier to write expressive unit tests for our api endpoints. 

This also incorporates the changes that were in #5 since that sets up infrastructure for a volunteers controller and model (I'm assuming via scaffolding).

The work is separated into a couple of commits. [efd41dc ](https://github.com/Code-For-Chicago/Voma-backend/pull/6/commits/efd41dc48cbeea7e1e48d40ea643666481648ad1 )is focused on adding in rsync, while the [commit before that](https://github.com/Code-For-Chicago/Voma-backend/pull/6/commits/b572f3dedba82006b2b3a51feb4bf2da57269c9e) is the actual end point work.

**New stuff this implements**
[rspec ](https://rspec.info/)  is the big new addition. It will enable us to write unit tests in a expressive and easy to read way. There's an example in this PR in `spec/controllers/volunteers_controller_spec.rb` . The aim of that file is to try to describe what expectations we have for behavior when some action takes place. In this situation since I'm writing an end point to handle a `POST` for creating a `Volunteer` it attempts to complete that action and then checks for a couple of things to ensure that a successful create happened. It also shows a non-happy scenario where the client has not provided a required parameter for `Volunteer` objects.

**How do I run the automated tests?**
You'll need to install postgres on your system and create a test database. I plan on creating a ticket for work to dockerize this in the future to make this easier to work with, but for now we need to set this up manually. Here's how:

1. [Download postgres](https://www.postgresql.org/download/) here and install it
2. Create a database in postgres called `Voma-backend_test` you'll also want to create one called `Voma-backend_development` if you want to play around with the end point manually. Here are [windows centric instructions](https://phoenixnap.com/kb/postgres-create-database) for how to create a database and more [generic instructions](https://www.postgresqltutorial.com/postgresql-create-database/).
3. Migrate the test database by running this:
```
rake db:migrate && rake db:test:prepare
```
4. Run the rspec tests by running
```
bundle exec rspec spec/
```
If all goes well you should get some output like this:
```
~/projects/Voma-backend$ bundle exec rspec spec/

Finished in 0.03986 seconds (files took 0.4818 seconds to load)
2 examples, 0 failures

```
